File: data/objects/entity/stat.py
class Stats:
    def __init__(self, health: int = 1, speed: int = 1, attack_speed: int = 1, defense: int = 0, attack: int = 1, vitality: int = 1):
        self.health = health
        self.speed = speed
        self.attack_speed = attack_speed
        self.defense = defense
        self.attack = attack
        self.vitality = vitality
    
    def copy(self) -> 'Stats':
        return Stats(self.health, self.speed, self.attack_speed, self.defense, self.attack, self.vitality)
    
    def from_json(data: dict) -> 'Stats':
        return Stats(data.get('health', 1), data.get('speed', 1), data.get('attack_speed', 1), data.get('defense', 1), data.get('attack', 0), data.get('vitality', 1))

File: server/src/main/java/dev/creoii/chaos/util/stat/Stat.java
package dev.creoii.chaos.util.stat;

import com.badlogic.gdx.utils.JsonValue;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class Stat {
    private final Type type;
    private int base;
    private final List<ModifierEntry> modifiers = new ArrayList<>();

    public Stat(Type type, int base) {
        this.type = type;
        this.base = base;
    }

    public Type type() {
        return type;
    }

    public int base() {
        return base;
    }

    public void set(int value) {
        this.base = value;
    }

    public void addModifier(ModifierEntry modifierEntry) {
        modifiers.add(modifierEntry);
    }

    public void removeModifier(UUID uuid) {
        modifiers.removeIf(modifier -> modifier.uuid().equals(uuid));
    }

    public int value() {
        int result = base;
        for (ModifierEntry mod : modifiers) {
            switch (mod.operation()) {
                case ADD -> result += mod.amount();
                case SET -> result = mod.amount();
                case MULTIPLY -> result *= mod.amount();
            }
        }
        return Math.max(0, result);
    }

    @Override
    public String toString() {
        return String.valueOf(value());
    }

    public static Stat parse(JsonValue jsonValue, Type type) {
        if (jsonValue != null) {
            if (jsonValue.isNumber()) {
                return new Stat(type, jsonValue.asInt());
            } else if (jsonValue.isObject()) {
                Stat stat = new Stat(type, jsonValue.getInt("value"));
                jsonValue.get("entries").forEach(entryValue -> stat.addModifier(ModifierEntry.parse(entryValue)));
                return stat;
            }
        }
        return new Stat(type, 0);
    }

    public enum Type {
        HEALTH,
        SPEED,
        ATTACK_SPEED,
        DEFENSE,
        ATTACK,
        VITALITY
    }
}
